### PR TITLE
DVS states + dependencies

### PR DESCRIPTION
**Pull request by alexbleotu**
### What does this PR do?

Implements new states and required execution and utils functions to manage distributed virtual switches (DVS) and distributed virtual portgroups (DVportgroup).

A new function file was added `salt.states.dvs`; in it's sysdoc there are examples of expected input. 

State functions:

- salt.states.dvs_configured state that configures/adds a DVS
- salt.states.portgroups_configured state that configures/adds/removes DVPortgroups
- salt.states.uplink_portgroup_configured state that configures the uplink portgroup of a DVS

Execution functions:

- salt.modules.vsphere.list_dvss to list dict representations of a DVS
- salt.modules.vsphere.create_dvs to create a DVS based on a dict representations
- salt.modules.vsphere.update_dvs to update a DVS based on a dict representations
- salt.modules.vsphere.list_dvportgroups to list dict representations of a DVPortgroups
- salt.modules.vsphere.list_uplink_dvportgroup to list the dict representation of the uplink portgroup of a DVS
- salt.modules.vsphere.create_dvportgroup to create a DVPortgroup based on a dict representations
- salt.modules.vsphere.update_dvportgroup to update a DVPortgroup based on a dict representations
- salt.modules.vsphere.remove_dvportgroup to remove a DVPortgroup

Utils functions: 

- salt.utils.vmware.get_dvss that retrieves DVSs in a datacenter
- salt.utils.vmware.get_network_folder that retrieves the network folder
- salt.utils.vmware.create_dvs
- salt.utils.vmware.update_dvs
- salt.utils.vmware.set_dvs_network_resource_management_enabled
- salt.utils.vmware.get_dvportgroups to retrieve distributed virtual portgroups
- salt.utils.vmware.get_uplink_dvportgroup to retrieve the uplink distributed virtual portgroup
- salt.utils.vmware.create_dvportgroup to create a distributed virtual portgroup
- salt.utils.vmware.update_dvportgroup to update a distributed virtual portgroup
- salt.utils.vmware.remove_dvportgroup

### Tests written?

Yes (utils functions only)

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by cachedout** [comment]
@alexbleotu Wow this is really great! While we get started on the reviews, could you please clean up the lint errors? They are here:

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/14747/violations/

This is going to be a fantastic addition!

**Comment by alexbleotu** [comment]
@cachedout had to rebase, there were some exceptions added in the PR you merged yesterday I needed in this one

**Comment by alexbleotu** [comment]
Don't know how to fix the remaining pylint error. Anyone know?

**Comment by cachedout** [comment]
Hey @rallytime I don't have my new laptop set up with pylint yet. Could you check on this error for @alexbleotu please?

**Comment by rallytime** [comment]
Yeah! @alexbleotu Your state module is using `range` so you need to use the one from the `six` lib included with salt. You can do that by simply adding `from salt.ext.six.moves import range` to your list of imports.

The Python 3 lint errors are not very helpful in the normal violation display, but you can find them in the console output of the lint build. (Totally not intuitive, I know, but works for now. We need to figure out what that output doesn't transfer to the jenkins error display.)

**Comment by alexbleotu** [comment]
Ah cool @rallytime! Thanks for the info! Will fix ASAP

**Comment by cachedout** [comment]
I didn't actually know that. Thanks, @rallytime !

**Comment by alexbleotu** [comment]
@cachedout Finally pylint done and all tests relevant to the change pass

File: salt/modules/vsphere.py [hunk 1]
```diff
@@ -3601,6 +3601,991 @@ def vsan_enable(host, username, password, protocol=None, port=None, host_names=N
...
+        List of DVS names to look for. If None, all DVSs are returned.
+        Default value is None.
+
+    .. code-block:: bash
+
+        salt '*' vsphere.list_dvss
+
+        salt '*' vsphere.list_dvss dvs_names=[dvs1,dvs2]
+    '''
+    ret_dict = []
```
Comment (cachedout) [inline salt/modules/vsphere.py]: `ret_dict` is a list? This is a bit confusing. :]

File: salt/modules/vsphere.py [hunk 2]
```diff
@@ -3601,6 +3601,991 @@ def vsan_enable(host, username, password, protocol=None, port=None, host_names=N
...
+def _apply_dvs_infrastructure_traffic_resources(infra_traffic_resources,
+                                                resource_dicts):
+    '''
+    Applies the values of the resource dictionaries to infra traffic resources,
+    creating the infra traffic resource if required
+    (vim.DistributedVirtualSwitchProductSpec)
+    '''
+    for res_dict in resource_dicts:
+        ress = [r for r in infra_traffic_resources if r.key == res_dict['key']]
+        if ress:
```
Comment (cachedout) [inline salt/modules/vsphere.py]: IMHO `ress` and `res` as variable names is very hard to read.

File: salt/states/dvs.py [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+        if __opts__['test']:
+            ret.update({'pchanges': changes,
+                        'result': None})
+        else:
+            ret.update({'changes': changes,
+                        'result': True})
+    return ret
+
+
+def _get_diff_dict(dict1, dict2):
```
Comment (cachedout) [inline salt/states/dvs.py]: Would it be better to move these private dictionary difference functions over to `salt.utils.dictdiffer`?

File: salt/modules/vsphere.py (hunk 1 above)
Comment (alexbleotu) [reply salt/modules/vsphere.py]: Old incarnation of the code. Will fix

File: salt/states/dvs.py (hunk 3 above)
Comment (alexbleotu) [reply salt/states/dvs.py]: Yep, I already created a `RecursiveDictDiffer` in `salt.utils.dictdiffer` which should be used here (it's in a different pull request). This is an pretty old code however and I didn't have time to update it. It's not such a trivial change. That's why I've upstreamed this code and will revisit later.

File: salt/modules/vsphere.py (hunk 2 above)
Comment (alexbleotu) [reply salt/modules/vsphere.py]: Will fix

**Review by cachedout — APPROVED** [review]

**Review by rallytime — APPROVED** [review]

**Comment by alexbleotu** [comment]
@cachedout rebased and added the pyVmomi incompatibility check

**Comment by alexbleotu** [comment]
Somehow some failing unit tests appeared due to https://github.com/saltstack/salt/pull/43645. Will fix ASAP

**Comment by alexbleotu** [comment]
some more test fixes

**Comment by rallytime** [comment]
Thank you very much for fixing up those tests, @alexbleotu! I guess we missed those failures when merging the other PR. I have fixed the boto and openscap test failures in other PRs.

**Pull request merged**